projects/hathor-wallet-mobile/0002-nano-contract-integration.md

```diff
@@ -0,0 +1,476 @@
+
+Upon entering a valid *Nano Contract ID* and tapping *Register*, users receive immediate feedback based on the action's outcome: loading, success, or failure.
+
+Upon successful registration, users are redirected to the *Main* screen  the with *Nano Contract List* active.
+
+##### Design Bill
+- **Placement**:
+	- Place the form input on the top of the screen.
+	- Place the action button on the bottom of the screen.
+
+- **Functionality**:
+	- Upon tapping *Register*, initiate the registration process with appropriate loading feedback.
+
+- **Visibility Conditions**:
+	- The *Register* button remains inactive until a valid Nano Contract ID is entered into the input field.
+	- In case of processing success, display a success feedback and navigate back to the *Main* screen with the *Nano Contract List* active.
+
+- **Error Management**:
+	- In case of validation error, display an error message near the input field *Nano Contract ID*.
+	- In case of processing failure, display an error feedback.
+
+## Wallet Connect Interaction
+
+The Wallet Connect has a different *modus operandi*, it requires the user to interact with the dapp by modals.
+
+We may want to make the Wallet Connect requests from dapps assynchronous with push notification, for this read the alternative [Push notification for wallet connect requests](#push-notification-wallet-connect-requests).
+
+```mermaid
+graph LR
+    A[Any Screen] --> B[Nano Contract Transaction Request, Modal]
+```
+
+### Modal: Nano Contract Transaction Request
+[modal-nano-contract-transaction-request](#modal-nano-contract-transaction-request)
+
+This modal presents detailed information for a *Nano Contract Transaction Request* from dapp, allowing the user to approve or reject the transaction. It activates when a dapp makes a `htr_createNcTx` request to the wallet's RPC.
+
+The modal presents detailed information containing *Blueprint ID*, *Blueprint Method*, *Caller*, *Nano Contract ID* (if applicable), *Action List*, and *Argument List*. An *Action List* can be either types *deposit* or *withdrawal*. Both the lists can be empty.
+
+##### Design Bill
+- **Placement**:
+	- The modal must render centrally on top of the current screen, ensuring clear visibility and focus.
+	
+- **Functionality**:
+	- Include detailed transaction information for user review:
+		- *Blueprint ID*, *Blueprint Method*, *Caller*, *Nano Contract ID* (if applicable), *Action List*, and *Argument List*.
+	- Include detailed information for each *Action List* item:
+		- *Type*, *Token* and *Data*.
+	- Include detailed information for *Data*:
+		- *Amount*, *Address*
+	- Include detailed information for each *Argument List* item:
+		- *Type* and *Value*
+	- Provide clear options for user approval or rejection of the transaction.
+	- Upon tapping *Approve*, initiate the processing with appropriate loading feedback.
```

Comment:
Since this might appear on any screen at any time, should we add a secondary confirmation modal to make sure the user intentionally clicked approve? I feel that he might misclick the button during normal usage of the app

Reply:
I don't like second modals, but we can use another strategy like tap in in a text "Take action on this request" and after it the buttons shows up. Change the action order. Set the primary button to "Reject" or "Dismiss".

Reply:
Other valid option is an Error Recovery strategy, such as undoing actions or restoring previous states.
This kind of strategy provides users with a clear and easy way to recover from errors, increasing the app safety.
If its something possible, we should consider as well: having only the reject or dismiss buttons, but an "undo previous action" message after.

Reply:
An undo button wouldn't work here because after the signed data is sent to the dApp, there is no going back

I like @alexruzenhack's suggestion tho